### PR TITLE
Accept any string as a key for `m.direct` account data

### PR DESCRIPTION
**Pull request by MatMaul**
This is the follow up of this [Ruma PR](https://github.com/ruma/ruma/pull/1946) for the SDK.

Ruma PR needs to be merged first.

### CODE REVIEW

**Review by bnjbvr — CHANGES_REQUESTED** [review]
I don't think losing static typing is an option, instead we should introduce an enum that can be one of all the types we expect (owned user id OR email address, if I understand correctly).

Also, `BaseRoomInfo` is serialized in the database, so we'd need a test making sure that we don't need a migration there, when the previous content was serialized as a `OwneduserId`; if the test fails, we need a migration.

**Comment by MatMaul** [comment]
> I don't think losing static typing is an option, instead we should introduce an enum that can be one of all the types we expect (owned user id OR email address, if I understand correctly).

Fine for me, but it also needs to support phone numbers I think, basically any [identifier types](https://spec.matrix.org/v1.12/client-server-api/#identifier-types).

Fine for you ?

**Comment by MatMaul** [comment]
Beware that it will not solve some the trouble for some legacy [buggy behavior](https://github.com/element-hq/element-x-android/issues/3074) of Element Web.

**Comment by bnjbvr** [comment]
> Fine for you ?

Fine if we do have some type checking / upfront validation for all these other id types indeed :)

**Comment by MatMaul** [comment]
Is there a way to not fully failed `m.direct` parsing when only some entries are invalid ?

The entry should just be committed in this case I believe.

Should I keep implementing `FromIterator<(String, Vec<OwnedRoomId>)>`, try to `.from` the string with the `UserIdentifier` enum we are talking about, and then put that in a `BTreeMap<UserIdentifier, Vec<OwnedRoomId>>` ?

**Comment by MatMaul** [comment]
So I think the question is, will serde uses `FromIterator<(String, Vec<OwnedRoomId>)>` or not to deserialize the JSON in this case ?

**Comment by spaetz** [comment]
If introducing ENUMS can we have an "UNKNOWN" too please which could catch either nonsense, such as "[object Object]": or other id's that this SDK perhaps does not know about (yet)?

**Comment by MatMaul** [comment]
Yes I think I have a way with something like that @spaetz , thanks.

File: Cargo.toml [hunk 1]
```diff
@@ -45,7 +45,7 @@ once_cell = "1.16.0"
 pin-project-lite = "0.2.9"
 rand = "0.8.5"
 reqwest = { version = "0.12.4", default-features = false }
-ruma = { version = "0.11.1", features = [
+ruma = { git = "https://github.com/ruma/ruma.git", rev = "35fda7f2f7811156df2e60b223dbf136fc143bc8", features = [
```
Comment (jplatte) [inline Cargo.toml:48]: Potentially stating the obvious here, but if you wanna do a crates.io release soon (I think this was planned), make sure to only merge this PR afterwards 😉

Comment (jplatte) [reply Cargo.toml:48]: Oh. Well I don't think we'll have the Ruma change blocking this PR out on crates.io anytime soon.

Comment (MatMaul) [reply Cargo.toml:48]: No rush in my side, it can wait.

Feel free to review in the meantime, so we can merge when we got a ruma release out.

Comment (poljar) [reply Cargo.toml:48]: Alright, the policy changed back, could you please rebase to pull in #4371, this will get rid of the CI error.

**Review by bnjbvr — COMMENTED** [review]
Does OwnedDirectUserIdentifier deserialize cleanly from an OwnedUserId? Otherwise this needs a migration, since we're storing the BaseRoomInfo in db

**Comment by bnjbvr** [comment]
> Does OwnedDirectUserIdentifier deserialize cleanly from an OwnedUserId? Otherwise this needs a migration, since we're storing the BaseRoomInfo in db

GOOD NEWS, EVERYONE! Since both are encoded as strings, and use the `IdZst` derive macro, they serialize and deserialize as a string, so this shouldn't require a migration. I would be grateful if a tiny test were added for this, to confirm my understanding: it would serialize an `OwnedUserId` and deserialize it as a `OwnedDirectUserIdentifier` cleanly. With that and the other review comment I've made above, LGTM :+1:

**Comment by MatMaul** [comment]
> I would be grateful if a tiny test were added for this

This feels like something that should be in Ruma, [here is the PR.](https://github.com/ruma/ruma/pull/1968)

File: crates/matrix-sdk-base/src/store/migration_helpers.rs [hunk 2]
```diff
@@ -200,12 +201,17 @@ impl BaseRoomInfoV1 {
             MinimalStateEvent::Redacted(ev) => MinimalStateEvent::Redacted(ev),
         });
 
+        let mut dm_targets_converted = HashSet::new();
+        for dm_target in dm_targets {
+            dm_targets_converted.insert(OwnedDirectUserIdentifier::from(dm_target));
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/store/migration_helpers.rs:206]:
```suggestion
        let mut converted_dm_targets = HashSet::new();
        for dm_target in dm_targets {
            converted_dm_targets.insert(OwnedDirectUserIdentifier::from(dm_target));
```

**Review by bnjbvr — APPROVED** [review]
lgtm, thanks.

**Pull request merged**